crates/matrix-sdk-base/src/sliding_sync.rs

```diff
@@ -305,15 +308,55 @@ impl BaseClient {
             let room = store.get_or_create_room(room_id, RoomState::Joined).await;
             let mut room_info = room.clone_info();
 
-            // We default to considering this room joined if it's not an invite, but we
-            // expect to find a m.room.member event in the required_state, so
-            // this should get fixed to the real correct value when we call
-            // self.handle_state below.
+            // We default to considering this room joined if it's not an invite. If it's
+            // actually left (and we remembered to request membership events in
+            // our sync request), then we can find this out from the events in
+            // required_state by calling handle_own_room_membership.
             room_info.mark_as_joined();
 
+            // We don't need to do this in a v2 sync, because the membership of a room can
+            // be figured out by whether the room is in the "join", "leave" etc.
+            // property. In sliding sync we only have invite_state and
+            // required_state, so we must process required_state looking for
+            // relevant membership events.
+            self.handle_own_room_membership(&room_data.required_state, &mut room_info).await;
+
             (room, room_info, None)
         }
     }
+
+    /// Find any m.room.member events that refer to the current user, and update the state in
+    /// room_info to reflect the "membership" property.
+    pub(crate) async fn handle_own_room_membership(
+        &self,
+        required_state: &[Raw<AnySyncStateEvent>],
+        room_info: &mut RoomInfo,
+    ) {
+        // Note: we deserialise these same events inside handle_state(). We could refactor to do
+        // this only once.
+
+        for raw_event in required_state {
+            let event = match raw_event.deserialize() {
```

Comment:
cc @jplatte: Is there an optimisation we can do here?

Reply:
Yes, we could do an `if raw_event.get_field::<StateEventType>("type").is_ok_and(|t| t == Some(StateEventType::RoomMember))` and if we enter that branch, deserialize using `raw_event.deserialize_as::<SyncRoomMemberEvent>()`. That way we never deserialize non-member events in this function.

Reply:
Sorry, this was confusing: my second commit avoids double-deserialising by doing it outside of this function.

If you're happy with that, we don't need to do anything else I think.